### PR TITLE
Follow symlinks in readdir() and readdirSync()

### LINKED ISSUE #14: `readdir` does not follow symlinks
```
fs = require('fs');
mockFs = require('mock-fs');
mockFs({
    '/fake': {
        'bar': {
            'hello.txt': 'Hello world'
        },
        'sym': mockFs.symlink({
            path: 'bar'
        })
    }
});

fs.readdir('/fake/bar', function(err, files) {
    console.log(err, files);
    fs.readdir('/fake/sym', function(err, files) {
        console.log(err, files);
    });
});
```

Expected output:
```
null [ 'hello.txt' ]
null [ 'hello.txt' ]
```

Actual output:
```
null [ 'hello.txt' ]
{ [Error: ENOTDIR, not a directory '/fake/sym']
  message: 'ENOTDIR, not a directory \'/fake/sym\'',
  code: 'ENOTDIR',
  errno: 27 } undefined
```

### PR DESCRIPTION
**Pull request by caitp**
Closes #14

### CODE REVIEW

**Comment by caitp** [comment]
Let me know if the indentation style used in tests aren't to your linking, not sure if the linter rules will complain if they're not.

This patch will resolve symlinks in the same way `open()` and `stat()` will, so I assume it's correct.

**Comment by caitp** [comment]
This isn't quite ready yet

**Comment by caitp** [comment]
This feels okay now

File: lib/filesystem.js [hunk 1]
```diff
@@ -73,9 +73,16 @@ FileSystem.prototype.getItem = function(filepath) {
...
   var currentParts = getPathParts(process.cwd());
   var item = this._root;
+  var itemPath = '/';
+  var linkPath;
   var name;
   for (var i = 0, ii = parts.length; i < ii; ++i) {
     name = parts[i];
+    if (item instanceof SymbolicLink) {
+      // Symbolic link being traversed as a directory
+      linkPath = path.resolve(path.dirname(itemPath), item.getPath());
```
Comment (caitp) [inline lib/filesystem.js]: there might be an issue here with `<link> -> <link> -> <file>`-type paths or something

**Comment by tschaub** [comment]
Thanks for this @caitp!  It looks like there might be more work needed to recursively follow links, but I think this is a nice enhancement.

**Comment by tschaub** [comment]
Published as `mock-fs@3.1.0`.

**Comment by caitp** [comment]
> Thanks for this @caitp! It looks like there might be more work needed to recursively follow links, but I think this is a nice enhancement.

I sent a followup with a bit more work on this portion --- I believe this matches the typical posix/libc behaviour, but I need to check that as it's been a while

**Pull request merged**